### PR TITLE
adds dynamic certificate reloading for kube aggregator

### LINKED ISSUE #87766: Dynamic reloading of certs in kube-aggregator
**What would you like to be added**:
1.17 made almost all certs in api-machinery world dynamically loaded from disk (example #84200) but the same was not done for certs in kube-aggregator. 

**Why is this needed**:
Allows e.g. cert rotation

/sig api-machinery
/milestone v1.18

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>

/kind feature

**What this PR does / why we need it**: 1.17 made almost all certs in `api-machinery` world dynamically loaded from disk but the same was not done for certs in `kube-aggregator`. This PR closes the gap by adding the code for dynamically loading certificates for the aggregator. 

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubernetes/issues/87766

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kube-aggregator certificates are dynamically loaded on change from disk

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -214,14 +212,30 @@ func (c completedConfig) NewWithDelegate(delegationTarget genericapiserver.Deleg
 	s.GenericAPIServer.Handler.NonGoRestfulMux.UnlistedHandle("/apis/", apisHandler)
 
 	apiserviceRegistrationController := NewAPIServiceRegistrationController(informerFactory.Apiregistration().V1().APIServices(), s)
+	if len(c.ExtraConfig.ProxyClientCertFile) > 0 && len(c.ExtraConfig.ProxyClientKeyFile) > 0 {
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:215]: we need to fail if the files aren't specified. That would match previous behavior, right?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -309,12 +323,10 @@ func (s *APIAggregator) AddAPIService(apiService *v1.APIService) error {
 
 	// register the proxy handler
 	proxyHandler := &proxyHandler{
-		localDelegate:   s.delegateHandler,
-		proxyClientCert: s.proxyClientCert,
-		proxyClientKey:  s.proxyClientKey,
-		proxyTransport:  s.proxyTransport,
-		serviceResolver: s.serviceResolver,
-		egressSelector:  s.egressSelector,
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:317]: this change surprises me.  what's going on?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go [hunk 3]
```diff
@@ -189,5 +189,17 @@ func (c *APIServiceRegistrationController) deleteAPIService(obj interface{}) {
 		}
 	}
 	klog.V(4).Infof("Deleting %q", castObj.Name)
-	c.enqueue(castObj)
+	c.enqueueInternal(castObj)
+}
+
+// Enqueue queues all apiservices to be rehandled.
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go:195]: this conforms to an interface? You'll want explain why.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go (hunk 2 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:317]: what exactly? we removed `proxyClientCert` and `proxyClientKey` but we added `proxyCurrentCertKeyContent`

Comment (p0lyn0mial) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:317]: indeed, it was missing - thx.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go (hunk 3 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go:198]: done.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go (hunk 1 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:216]: I identified only https://github.com/kubernetes/kubernetes/pull/92791/files#diff-c070121c775ca8660b40b691e24bc6c4L156 as a potential place that could have failed when the files were missing.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 4]
```diff
@@ -183,6 +153,34 @@ func (c *AvailableConditionController) sync(key string) error {
...
+		},
+	}
+
+	if c.proxyCurrentCertKeyContent != nil {
+		proxyClientCert, proxyClientKey := c.proxyCurrentCertKeyContent()
+
+		restConfig.TLSClientConfig.CertData = proxyClientCert
+		restConfig.TLSClientConfig.KeyData = proxyClientKey
+	}
+	if c.proxyTransport != nil && c.proxyTransport.DialContext != nil {
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:171]: you lost the egress proxy stanza. Please restore it.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 5]
```diff
@@ -183,6 +153,34 @@ func (c *AvailableConditionController) sync(key string) error {
...
+		restConfig.Dial = c.proxyTransport.DialContext
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:172]: If all you're using is the DialContext, please only store that in the `AvailableController` struct

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 6]
```diff
@@ -110,39 +111,8 @@ func NewAvailableConditionController(
...
-	// that's not so bad) and sets a very short timeout.  This is a best effort GET that provides no additional information
-	restConfig := &rest.Config{
-		TLSClientConfig: rest.TLSClientConfig{
-			Insecure: true,
-			CertData: proxyClientCert,
-			KeyData:  proxyClientKey,
-		},
-	}
-
-	if egressSelector != nil {
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:126]: you lost this

**Comment by deads2k** [comment]
The unit test is good. I'd like to see an integration test please.

**Comment by fedebongio** [comment]
Thank you David!

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go (hunk 4 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:171]: thanks.

**Comment by deads2k** [comment]
this is the last cert/key pair that isn't dynamic.  It doesn't match user expectations to have one odd man out.  This PR fixes it.

/kind bug

**Pull request merged**